File: CS 3450 Code Examples/3450Code/Restaurant.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
using namespace std;

class Cook {
public:
    void makeBurger() {
        cout << "Making a burger\n";
    }
    void makeFries() {
        cout <<"Making fries\n";
    }
    void makeShortStack() {
        cout <<"Making 2 pancakes\n";
    }
    void makeScrambled() {
        cout <<"Scrambling eggs\n";
    }
    void makeHashBrowns() {
        cout <<"Making hash brown potatoes\n";
    }
} cook;


template<class Order>
void orderUp(Order& order) {
    order();
}


class BurgerAndFries {
    Cook& receiver;
public:
    BurgerAndFries(Cook& c) : receiver(c) {}
    void operator()(){
        receiver.makeBurger();
        receiver.makeFries();
    }
};

class FarmersSpecial {
    Cook& receiver;
public:
    FarmersSpecial(Cook& c) : receiver(c) {}
    void operator()(){
        receiver.makeShortStack();
        receiver.makeScrambled();
        receiver.makeHashBrowns();
    }
};

class Waitress {
public:
    static void serve() {
		BurgerAndFries baf(cook);
		FarmersSpecial f(cook);
        for (int i = 0; i < 5; ++i) {
            // Take order and serve
            switch (rand() % 2) {
            case 0:
                orderUp(baf);
                break;
            case 1:
				orderUp(f);
                break;
            }
            cout << endl;
        }
    }
};

int main() {
    srand(time(0));
    Waitress::serve();
}

File: CS 3450 Code Examples/JavaAll/Restaurant.java
import java.util.Random;

// The Command
interface Order {
    void orderUp();
}

class Cook {
    public void makeBurger() {
        System.out.println("Making a burger");
    }
    public void makeFries() {
        System.out.println("Making fries");
    }
    public void makeShortStack() {
        System.out.println("Making 2 pancakes");
    }
    public void makeScrambled() {
        System.out.println("Scrambling eggs");
    }
    public void makeHashBrowns() {
        System.out.println("Making hash brown potatoes");
    }
}

class BurgerAndFries implements Order {
    private Cook receiver;
    public BurgerAndFries(Cook c) {
        receiver = c;
    }
    public void orderUp(){
        receiver.makeBurger();
        receiver.makeFries();
    }
}

class FarmersSpecial implements Order {
    private Cook receiver;
    public FarmersSpecial(Cook c) {
        receiver = c;
    }
    public void orderUp() {
        receiver.makeShortStack();
        receiver.makeScrambled();
        receiver.makeHashBrowns();
    }
}

class Waitress {
    public static void serve() {
        Random ran = new Random();
        Order order = null;
        for (int i = 0; i < 8; ++i) {
            // Take order
            switch (ran.nextInt(2)) {
            case 0:
                order = new BurgerAndFries(Restaurant.cook);
                break;
            case 1:
                order = new FarmersSpecial(Restaurant.cook);
                break;
            }

            // Submit and process order (the command)
            order.orderUp();
            System.out.println();
        }
    }
}

class Restaurant {
    static Cook cook = new Cook();
    public static void main(String [] args) {
        Waitress.serve();
    }
}
